### PR TITLE
[composite compliance] ctc_loss

### PR DESCRIPTION
**Pull request by kshitij12345**
#Ref #69991

I have mixed feelings about adding new (private) operators. Backends writers will have to override them as well.

### CODE REVIEW

**Comment by kshitij12345** [comment]
ASAN Failure: (Seems like this existed already, this PR just caught it)

https://github.com/pytorch/pytorch/runs/8307959983?check_suite_focus=true#step:10:1671

File: test/test_proxy_tensor.py [hunk 1]
```diff
@@ -1182,6 +1180,7 @@ def f(a, b, c, d, e):
     xfail('nn.functional.cosine_embedding_loss', ''),  # The underlying op of 'aten.stride' has no overload name '_schema'
     xfail('nn.functional.cosine_similarity', ''),  # aten.size.default - couldn't find symbolic meta function/decomposition
     xfail('nn.functional.cross_entropy', ''),  # aten.size.default - couldn't find symbolic meta function/decomposition
+    xfail('nn.functional.ctc_loss'),  # aten._ctc_loss.Tensor - couldn't find symbolic meta function/decomposition
```
Comment (kshitij12345) [inline test/test_proxy_tensor.py:1183]: Not a regression

**Comment by kshitij12345** [comment]
@zou3519 what do you think of this approach?
Thanks!

**Comment by zou3519** [comment]
> I have mixed feelings about adding new (private) operators. Backends writers will have to override them as well.

The final end state of this should be that we have 4 operators:
- ctc_loss, ctc_loss_backward
- ctc_loss.Tensor, ctc_loss_backward.Tensor

I would expect there to be some way to unify the `_cudnn_ctc_loss` with the regular `_ctc_loss`, just like how we unified all of the convolutions into two operators (convolution and convolution_backward).

Your PR is on the right track, though it also adds additional .Tensor overloads for the cudnn_ctc_loss* operators. I see "unifying the _cudnn_ctc_loss with the regular _ctc_loss" as orthogonal, so this PR LGTM.

**Review by zou3519 — APPROVED** [review]

**Comment by ezyang** [comment]
maybe @eellison can help with the asan error under fake tensor

**Comment by kshitij12345** [comment]
I am planning to merge this. Will assign @eellison to the relevant issue (#85231)

File: aten/src/ATen/native/LossCTC.cpp [hunk 2]
```diff
@@ -397,17 +447,30 @@ Tensor ctc_loss(const Tensor& log_probs_, const Tensor& targets, IntArrayRef inp
...
+} // namespace
+
+Tensor ctc_loss(const Tensor& log_probs_, const Tensor& targets, IntArrayRef input_lengths, IntArrayRef target_lengths, int64_t BLANK, int64_t reduction, bool zero_infinity) {
+  return ctc_loss_impl(log_probs_, targets, input_lengths, target_lengths, BLANK, reduction, zero_infinity);
+}
+
 // Convenience function accepting Tensors
 Tensor ctc_loss(const Tensor& log_probs, const Tensor& targets, const Tensor& input_lengths, const Tensor& target_lengths, int64_t BLANK, int64_t reduction, bool zero_infinity) {
+  if (at::areAnyTensorSubclassLike(
+          {log_probs, targets, input_lengths, target_lengths})) {
```
Comment (bdhirsh) [inline aten/src/ATen/native/LossCTC.cpp:467]: Hey @zou3519 - `ctc_loss()` is an example of a "non-composite compliant decomposition", where it looks like we work around it by adding a composite compliant path that only executes some of the time.

Here, it looks like we only run it when any inputs are tensor subclasses. But, I think we want this path to also active wait:
- any transforms are active (specifically functionalization, and probably the other functorch transforms
- any python modes are active (maybe that's also checked by `areAnyTensorSubclassLike`)


Do you agree with that? Should we update `areAnyTensorSubclassLike()` (and/or rename it) to properly reflect that criteria?

That should fix https://github.com/pytorch/pytorch/issues/86384

Comment (zou3519) [reply aten/src/ATen/native/LossCTC.cpp:467]: I thought this path already is active on functorch transforms, but I am not sure about the interaction with python modes. IMO `areAnyTensorSubclassLike` should cover all functorch transforms and Python Modes

Comment (bdhirsh) [reply aten/src/ATen/native/LossCTC.cpp:467]: Oh maybe this is the bug. This keyset should probably include `DispatchKey::Functionalize`, right? I can make a patch

https://github.com/pytorch/pytorch/blob/bc919ac7963be9f113b4e3c8b668905404301f8f/aten/src/ATen/TensorSubclassLikeUtils.h#L27

Comment (zou3519) [reply aten/src/ATen/native/LossCTC.cpp:467]: Yeah that looks like the bug, good find

**Pull request closed without merging**